app/src/main/java/fr/gaulupeau/apps/Poche/data/StorageHelper.java

```diff
@@ -55,42 +55,78 @@ public static String readRawStringUnsafe(int id) throws IOException {
         }
     }
 
-    public static String getExternalStoragePath() {
-        if(externalStoragePath == null) {
-            String returnPath = null;
+    public static String getFirstExternalStoragePath()
+    {
+        File[] externalFilesDirs = ContextCompat.getExternalFilesDirs(App.getInstance(), null);
+        for(File extStorageDir: externalFilesDirs) {
+            if(extStorageDir == null) {
+                continue;
+            }
+
+            return extStorageDir.getPath();
+        }
+        return null;
+    }
+
+    public static String getExternalStoragePath()
+    {
+        String storagePathSetting = App.getSettings().getDbPath();
+
+        if(storagePathSetting == "external") {
             File[] externalFilesDirs = ContextCompat.getExternalFilesDirs(App.getInstance(), null);
-            // TODO: better SD Card detection
             for(File extStorageDir: externalFilesDirs) {
                 if(extStorageDir == null) {
-                    Log.w(TAG, "getExternalStoragePath() extStorageDir is null");
                     continue;
                 }
 
-                returnPath = extStorageDir.getPath();
-                Log.d(TAG, "getExternalStoragePath() extStorageDir.getPath(): " + returnPath);
-                break;
+                return extStorageDir.getPath();
+            }
+        }
+        return storagePathSetting;
+    }
+
+    public static String[] getExternalStoragePaths() {
+        String returnPath = null;
+        File[] externalFilesDirs = ContextCompat.getExternalFilesDirs(App.getInstance(), null);
+        ArrayList<String> usableExternalFilesDirs = new ArrayList<String>();
+        for(File extStorageDir: externalFilesDirs) {
+            if(extStorageDir == null) {
+                Log.w(TAG, "getExternalStoragePaths() extStorageDir is null");
+                continue;
             }
 
-            Log.d(TAG, "getExternalStoragePath() returnPath: " + returnPath);
-            return externalStoragePath = returnPath;
+            usableExternalFilesDirs.add(extStorageDir.getPath());
+            Log.d(TAG, "getExternalStoragePaths() extStorageDir.getPath(): " + extStorageDir.getPath());
         }
 
-        return externalStoragePath;
+        String[] result = new String[usableExternalFilesDirs.size()];
+        result = usableExternalFilesDirs.toArray(result);
+        return result;
     }
 
     public static boolean isExternalStorageReadable() {
-        String externalStoragePath = getExternalStoragePath();
-        if(externalStoragePath == null) return false;
+        String storagePath = App.getSettings().getDbPath();
+        if(TextUtils.isEmpty(storagePath)) {
+            storagePath = getFirstExternalStoragePath();
+        }
+        if(storagePath == null) return false;
```

Comment:
maybe check with TextUtils.isEmpty also for empty strings?

Reply:
Assuming you mean the check on line 112, if the setting is not set, Settings.getDbPath() will return null (because that's what Settings.getString() does.  getFirstExternalStoragePath() also always returns a non-empty string, or null (it just returns results from ContextCompat.getExternalFilesDirs(), which returns null, or a list of String objects that aren't empty).  I don't think there's any way for the code to encounter an empty string here, unless the app for some reason stored a blank string for the dbPath setting.  If you want it to check anyways, I can change it, but I don't think it can ever encounter this scenario.